### ISSUE TITLE
Subgraph {iso|mono}morphism is very slow

### ISSUE THREAD

**Issue by j6k4m8** [networkx/networkx#4262]
The `algorithms.isomorphism.GraphMatcher#subgraph_{mono|iso}morphisms_iter` results appear to be very slow; I think there may be some unnecessary steps taking place here. Is this a known issue? Is there interest in maturing this part of the codebase?

**Comment by dschult** [comment]
Yes, there seems to be some active improvements in the literature in recent years.
I have a version of the VF2++ algorithm from a student of mine that I hope to include. 
But even that algorithm is 2 years old at this point... There may be improvements since then. 

See also [many Issues, both open and closed](https://github.com/networkx/networkx/search?q=vf2&type=issues)

**Comment by j6k4m8** [comment]
I'd also be willing to contribute [this implementation](https://github.com/aplbrain/grandiso-networkx), which probably needs a bit of love to get it into the networkx API but is quite comparatively performant!
<img width="485" alt="image" src="https://user-images.githubusercontent.com/693511/96184546-a35e0380-0f06-11eb-8475-1921e8f94256.png">

**Comment by dschult** [comment]
That performance looks very nice :)   We'd love to have it in NX!
And it's also nice that this is published... the code has more documentation and the paper gets more exposure too. 
If you have questions about how to get it into NetworkX please ask...  I'm motivated to get this in.
Also, you can create a PR before it's completely ready and get feedback along the way.

I'd put it in the ```isomorphism``` folder within ```algorithms```. 
Make a hook for it in the documentation: doc/reference 
Run "black" for setting up the code style. 
Use Sphinx numpydoc doc_strings (one-liner first line, then blank line, then paragraph describing what it does, then parameter defns and return values, examples, etc.) 
Look at existing modules to find examples.

More here:  https://networkx.github.io/documentation/stable/developer/contribute.html

**Comment by j6k4m8** [comment]
Very cool @dschult — would you be opposed to importing the library via a `requirements/default.txt` dependency rather than copying the full codebase over?

**Comment by dschult** [comment]
That's not very specific. Would there be code in NX to call this other library? Or would there be code in the other library to run it on a NX graph? 

I think the second choice is better for independence of the library -- and avoids getting it added as a dependency of networkx. We want to keep the number of dependencies down because many people want NX who don't need/want subgraph isomorphism. 

Also, separate libraries without a long track record tend to become unmaintained over time. 
It's probably best (if you want it separate) to make it a parallel library that can take a NetworkX Graph as input. You could include an example in the NX gallery which shows people how to use it.

**Comment by j6k4m8** [comment]
Got it. Was more asking if you’d like a full copy of the source added to nx
as a submodule vs a dependency import: you answered my question :)

The library already runs on a networkx graph and digraph, so whatever your
preference, happy to oblige!

**Comment by dschult** [comment]
We'd love to have it as part of NetworkX.
I would put it in the `networkx/algorithms/isomorphism` folder as a new module with tests in a module `.../tests/test_modulename.py`

**Comment by MridulS** [comment]
@j6k4m8 Just wanted to check in and see if you would like to contribute the code inside networkx or would like to keep it an independent library. Either is fine :)

BTW we have an implementation of VF2++, maybe you would like to benchmark against it too https://github.com/networkx/networkx/blob/main/networkx/algorithms/isomorphism/vf2pp.py

**Comment by j6k4m8** [comment]
awesome, so glad to see vf2++ is in the library now!! I will try some benchmarks and if we still seem competitive I'll write up a PR :) otherwise it doesn't seem like there's a great reason to have THREE subgraph iso algos in nx, unless you think there's an advantage?
